### PR TITLE
Fix stream sync logic to respect dependencies and add comprehensive unit tests

### PR DESCRIPTION
**Pull request by rsaha-qlik**
# Description of change
Fixes https://qlik-dev.atlassian.net/browse/SAC-30068

### Stream Sync Architecture
- **Fixed stream dependency handling**: Refactored sync logic to ensure child streams are synced through their parent streams
  - All list-dependent streams (messages, subscribed_contacts) now sync through `sync_lists()`
  - All message-dependent streams (message_clicks, message_opens, etc.) now sync through `sync_messages()`
- **Added schema loading**: Added `schemas.load_and_write_schema()` calls to all sync functions to ensure schemas are written before data
- **Made stream dependencies explicit**: Child streams now require parent stream data to be passed explicitly (e.g., `sync_subscribed_contacts` now requires lists parameter)

### Testing Infrastructure
- **Added comprehensive unit tests** (`tests/unittests/test_syncs.py`):
  - Tests for all sync functions with mocked responses
  - Tests verifying schema loading behavior
  - Tests verifying parent-child sync dependencies
  - Tests for error conditions and edge cases
- **Enhanced HTTP retry tests** (`tests/unittests/test_http.py`):
  - Added `mock_sleep` fixture to speed up retry tests
  - Added assertions to verify retry behavior
- **Updated CircleCI config** to run unit tests with coverage reporting

### Bug Fixes
- Fixed edge case where `None` response in `sync_lists` would cause errors (added `or []` fallback)
- Removed dead code in `sync_subscribed_contacts` that would fetch lists when not provided

## Testing
- All unit tests pass
- CircleCI pipeline includes unit test execution with coverage

# Rollback steps
- Revert this branch

#### AI generated code
- [x] This PR has been written with the help of GitHub Copilot or another generative AI tool

### CODE REVIEW

File: tests/unittests/test_syncs.py
Comment (RushiT0122) [inline tests/unittests/test_syncs.py:1]: Unit tests are not running in the CCi, you need to configure it in the CCi config file.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]

File: tests/unittests/test_syncs.py [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+        mock_request.return_value = [{'ListID': '1', 'Name': 'Test List'}]
+        
+        streams.sync_lists(self.ctx)
+        
+        # Assert that lists were fetched and written
+        mock_request.assert_called_once()
+        mock_write_records.assert_called_once()
+
+if __name__ == '__main__':
+    unittest.main()
```
Comment (RushiT0122) [inline tests/unittests/test_syncs.py]: Add last line.

Comment (rsaha-qlik) [reply tests/unittests/test_syncs.py]: Addressed

File: tests/unittests/test_syncs.py
Comment (rsaha-qlik) [reply tests/unittests/test_syncs.py:1]: added unittests in circleci now

File: tests/unittests/test_syncs.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+        streams._sync_messages(self.ctx, [{'ListID': '1', 'Name': 'Test List'}])
+
+        # Assert that the request was called for messages
+        self.assertEqual(mock_request.call_count, 1)
+
+        # Assert that write_records was called with the correct data
+        mock_write_records.assert_called_with(
+            streams.IDS.MESSAGES,
+            [{'MsgID': '1', 'SendDate': '2026-01-15T00:00:00Z'}]
+        )
```
Comment (RushiT0122) [inline tests/unittests/test_syncs.py:111]: Duplicate assertion.

File: tap_listrak/__init__.py [hunk 3]
```diff
@@ -50,19 +50,40 @@ def discover(ctx):
...
+        # For any other independent streams (if added in future)
+        for tap_stream_id in ctx.selected_stream_ids:
+            if tap_stream_id not in lists_dependent_streams:
+                schemas.load_and_write_schema(tap_stream_id)
+                if hasattr(streams_, f"sync_{tap_stream_id}"):
+                    sync_fn = getattr(streams_, f"sync_{tap_stream_id}")
+                    LOGGER.info(f"Syncing stream: {tap_stream_id}")
+                    sync_fn(ctx)
+                else:
+                    LOGGER.warning(f"No sync function found for stream: {tap_stream_id}")
```
Comment (RushiT0122) [inline tap_listrak/__init__.py]: This seems like dead code block. Since `lists` stream is by default selected, seems like execution is never going into else code block.

**Review by RushiT0122 — CHANGES_REQUESTED** [review]

File: tap_listrak/__init__.py [hunk 4]
```diff
@@ -71,19 +80,30 @@ def sync(ctx):
...
-                schemas.load_and_write_schema(tap_stream_id)
-                if hasattr(streams_, f"sync_{tap_stream_id}"):
-                    sync_fn = getattr(streams_, f"sync_{tap_stream_id}")
-                    LOGGER.info(f"Syncing stream: {tap_stream_id}")
-                    sync_fn(ctx)
-                else:
-                    LOGGER.warning(f"No sync function found for stream: {tap_stream_id}")
+
+    # In future we need to add streams independent of lists, 
+    # we can add their sync calls in `else` block.
```
Comment (RushiT0122) [inline tap_listrak/__init__.py]: We don't need this comment.

File: tap_listrak/__init__.py [hunk 5]
```diff
@@ -50,19 +59,51 @@ def discover(ctx):
...
-            LOGGER.warning(f"No sync function found for stream: {tap_stream_id}")
+    
+    # Check if any stream requires lists to be synced
+    lists_dependent_streams = {'lists', 'messages', 'subscribed_contacts', 
+                               'message_clicks', 'message_opens', 'message_reads',
+                               'message_sends', 'message_unsubs', 'message_bounces'}
+    
+    needs_lists_sync = any(stream in ctx.selected_stream_ids for stream in lists_dependent_streams)
+    
+    if needs_lists_sync:
```
Comment (RushiT0122) [inline tap_listrak/__init__.py]: IMO, we don't need this now.

File: CHANGELOG.md [hunk 6]
```diff
@@ -1,5 +1,8 @@
 # Changelog
 
+## 1.3.0
+  * Refactors stream sync logic to respect parent-child dependencies. Adds schema loading to all sync functions to ensure schemas are written before records. Adds comprehensive unit tests for sync functions, schemas, context, and HTTP retry logic. Updates CircleCI configuration to run unit tests with coverage reporting. Fixes edge case where None response in sync_lists could cause errors [#27](https://github.com/singer-io/tap-listrak/pull/27)
+
```
Comment (RushiT0122) [inline CHANGELOG.md]:
```suggestion
  * Refactors stream sync logic to respect parent-child dependencies. [#27](https://github.com/singer-io/tap-listrak/pull/27)
  * Fixes edge case where None response in sync_lists could cause errors
  * Adds schema loading to all sync functions to ensure schemas are written before records.
  * Adds comprehensive unit tests for sync functions, schemas, context, and HTTP retry logic.
  * Updates CircleCI configuration to run unit tests with coverage reporting.

```

**Review by RushiT0122 — APPROVED** [review]
